### DISCUSSION TITLE
Google Summer of Code 2023 application

### DISCUSSION THREAD

**Discussion by shaheemazmalmmd** [fossology/fossology#2371] [Ideas]
Hello all!

FOSSology as an org is planning to apply for [Google Summer of Code](https://summerofcode.withgoogle.com) this year as well. In preparation to do so, we are creating this discussion to gather ideas for the GSoC-2023 projects.

This year, there are some changes made by Google in the program. You can read more about it on [their website](https://opensource.googleblog.com/2021/11/expanding-google-summer-of-code-in-2022.html). But as a summary,
- Two sized projects are acceptable by program:
  + Medium sized projects (~175 hours)
  + Large projects (~350 hours)
- The program is open to 18 years and older people.
  + No longer limited to college students only.
- Flexible time lines between 12 and 22 weeks.

Please feel free to drop any questions here (or start a new discussion) you have regarding the program, you want to submit a proposal idea, you want to be mentor in the program, etc.

While submitting a project idea, please tag it to be "Medium" or "Large" sized and who you'll prefer working on it "student" or "professional".
Some resources from Google:
- https://google.github.io/gsocguides/mentor/defining-a-project-ideas-list
- https://google.github.io/gsocguides/mentor/making-your-ideas-page

You can also check the detailed documentation of previous students from [GSoC-2022](https://fossology.github.io/gsoc/docs/2022).

**Comment by GMishx** [comment]
# Idea!
## Title: REST API improvements
### Goal: Making all data from current UI available on REST API
- Writing test cases for all the existing and new functionalities. 
- Improve REST API and expose more endpoints
  - Update REST API to v2
  - https://github.com/fossology/fossology/issues/2404
  - https://github.com/fossology/fossology/issues?q=is%3Aissue+is%3Aopen+REST
  - https://github.com/Shruti3004/FOSSology-REST-API/issues

| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | - |
| Risk/Exploratory | * |
| Fun/Periphial | ** |
| Core Development | *** |
| Project Infrastructure | ** |
| Project size | Large |
| Preferred contributor | Student/Professional |

**Reply by viv9k** [reply]
I will participate in mentoring this.

**Reply by subhamengine** [reply]
I would like to contribute to this.

**Reply by bethelson** [reply]
I will like to contribute to this. Are there any pre-requisits or tasks to carry out concerning this before the application phase?

**Reply by mayank-pathakk** [reply]
@viv9k @Shruti3004 Who are the mentors to this project ?

**Reply by mayank-pathakk** [reply]
@bethelson Did you find something , are there some pre-requisites ?

**Reply by mayank-pathakk** [reply]
@shaheemazmalmmd  Could you tell me who are the mentors for this project ?.
Thanks in advance

**Reply by shaheemazmalmmd** [reply]
@mayank-pathakk you can contact @GMishx he will be one of the mentors.

**Comment by GMishx** [comment]
# Idea!
## Title: Integrating Open Source Review Toolkit
### Goal: Using ORT to fetch dependencies and generate SBOM

Build systems fetch the required dependencies (library/artifact) for a project while building the project. Its important to get an insight of these dependencies for license compliance check.

The [OSS Review Toolkit](https://github.com/oss-review-toolkit/ort) is an open source project helps to find dependencies in a project.

The goal of this project is to render the project dependencies created by ort and display those in the fossology-UI. Dependencies can be scheduled directly from the UI and scan with fossology.

| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | - |
| Risk/Exploratory | - |
| Fun/Periphial | ** |
| Core Development | *** |
| Project Infrastructure | * |
| Project size | Large |
| Preferred contributor | Student/Professional |

**Reply by GMishx** [reply]
Or use the other approach recommended by @sameer1046: https://github.com/fossology/fossology/discussions/2140#discussioncomment-2112774

**Comment by GMishx** [comment]
# Idea!
## Title: Reducing false positive copyrights
### Goal: Improving copyright agent by reducing false positives
In GSoC 2021, @Kaushl2208 worked on an ML script to tag and cleanup false positive copyrights detected by FOSSology. The goal is to continue on the work and improve the script. Alternatively, the copyright agent of FOSSology itself can be improved to achieve better results.
Known issues:
- The script currently works only for English names.
- The clutter removal changes the casing of text.
- Low accuracy of the agent overall.

Links:
- GSoC report: https://github.com/Kaushl2208/GSoC2021
- Pull request: https://github.com/fossology/fossology/pull/2083

| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | ** |
| Risk/Exploratory | *** |
| Fun/Periphial | ** |
| Core Development | * |
| Project Infrastructure | ** |
| Project size | Large |
| Preferred contributor | Student |

**Reply by Kaushl2208** [reply]
I would love to help with this. 

- Currently the defined model is pre-trained.
- Training NER model on custom dataset instead of general datasets will help in Efficiency and accuracy of the model.
- The format of POS tagging can be further enhanced to various permutations.

**Reply by hastagAB** [reply]
I would like to mentor for this project.

**Reply by jannatpritha** [reply]
Is there any prerequisite for this project ? And is there any demo for proposal letter?

**Reply by jannatpritha** [reply]
@Kaushl2208  The code is not available on the repo. I am not sure how should I contribute to this project. It will be helpful if u could help me out.  I do have some ideas that I want to present. Having some data from training and test data would help.

**Reply by Kaushl2208** [reply]
Hey @jannatpritha , I appreciate your enthuthiasm but if you have read the description you can find the link to a PR, Which contains all the details from files to integration :)
Your ideas are always welcome. I will look for the dataset which I worked on and will provide on #gsoc-2023 channel on slack.

Hope that helps

**Comment by GMishx** [comment]
# Idea!
## Title: Support CycloneDX report
### Goal: Creating CycloneDX reports along with SPDX and other formats
CycloneDX is another popular format for generating SBOMs. Having FOSSology generate reports in CycloneDX along with SPDX will enable integration with more tools.

Links:
- https://packagist.org/packages/cyclonedx/cyclonedx-library
- https://cyclonedx.org/

| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | ** |
| Risk/Exploratory | * |
| Fun/Periphial | ** |
| Core Development | * |
| Project Infrastructure | ** |
| Project size | Large |
| Preferred contributor | Student |

**Reply by GMishx** [reply]
As a small enhancement, other reporting formats from SPDX can be implemented as well including JSON and YAML.

**Reply by Kaushl2208** [reply]
This looks interesting!! Would love to be a part of this project :)

**Reply by jkowalleck** [reply]
Hi, I am the author and current maintainer of that https://packagist.org/packages/cyclonedx/cyclonedx-library 
If you need help, or just want to have additional documentation written or features and data models implemented, 
feel free to create a request/issue here:  https://github.com/CycloneDX/cyclonedx-php-library.

If you want to chat abut the CycloneDX topics, feel free to [join our Slack](https://cyclonedx.org/slack/invite) and write me in channel [CycloneDX#php-library](https://cyclonedx.slack.com/archives/C02E57HTBL3)


Please be aware, that by Summer 2023 the v2 of that library might be available, which has critical changes to the current v1. 
Details can already be found here: https://github.com/CycloneDX/cyclonedx-php-library/blob/master/HISTORY.md 
New code usage will be pretty similar to the one from V1, the documentation quality will slightly increase, and there will still be a `phpdoc3` config file to create your own docs if needed.

**Reply by shaheemazmalmmd** [reply]
Hello @jkowalleck,

Thank you for your suggestions and valuable information. We will contact you incase any further information required.

**Comment by shaheemazmalmmd** [comment]
# Idea!
## Title: Support conf file & UI for NOMOS agent
### Goal: Adding text phrases from UI or from .conf file should enble nomos to scan for new text phrases(Added) for licenses

FLOW : Create a UI Where user can add multiple text phrases associated with license(FROM FOSSology License Database)
       This text phrases should get converted in to regex and stored in a conf.
       Should provide ability to NOMOS to copy regex and associated license from conf to search for a license.
       Test cases needs to be provided as well.


| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | ** |
| Risk/Exploratory | * |
| Fun/Periphial | ** |
| Core Development | * |
| Project Infrastructure | ** |
| Project size | Large |
| Preferred contributor | Student/professional |

**Comment by shaheemazmalmmd** [comment]
# Idea!
## Title: Support text phrases scanning for MONK agent
### Goal: Adding text phrases from UI to database and provide ability to scan them using MONK.

FLOW : Create a UI Where user can add multiple text phrases associated with license(FROM FOSSology License Database)
       Should enable MONK to scan from phrases table.
       Test cases needs to be provided as well.


| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | ** |
| Risk/Exploratory | * |
| Fun/Periphial | ** |
| Core Development | * |
| Project Infrastructure | ** |
| Project size | Medium |
| Preferred contributor | Student/professional |

**Reply by viv9k** [reply]
I will participate in mentoring this.

**Reply by PranitPatil03** [reply]
I am interested in contributing to this project and would like to learn more about the Project.

**Comment by shaheemazmalmmd** [comment]
# Idea!
## Title: New project to estimate clearing effort and create a bucket
### Goal: Create a new project using bootstrap and PHP/GO/Python to estimate time to clear the given number of components.


FLOW : Create a UI Where user can upload CSV/JSON
       Read all the components in the given file and search them in FOSSology & SW360 Database.
       List the components that were already cleared in FOSSology AND sw360.
       List all the components that were cleared with other versions.
       List all the components that are not matched with any of the components.
       And also list a ESTIMATED TIME to finish the upload based upon size.
       Sync with FOSSology & SW360 And maintain the list of the projects in database with SBOM.
       Advanced search functionality to search for the available components mainly for developers.  


| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | * |
| Risk/Exploratory | * |
| Fun/Periphial | ** |
| Core Development | *** |
| Project Infrastructure | *** |
| Project size | Large |
| Preferred contributor | Student/professional |

**Reply by viv9k** [reply]
Is there any specific idea behind choosing php for new project, can we take in react and bootstrap 5?

**Reply by NicolasToussaint** [reply]
That's interesting ; the key point will be how to uniquely identify a given component, and to link it to existing upload in fossology

**Reply by Simran180** [reply]
I want to contribute to this project

**Reply by PranitPatil03** [reply]
I am interested in contributing to this project and would like to learn more about the Project.

**Comment by ag4ums** [comment]
# Idea!
## Title: Main license detection for components
### Goal: Automatically detect main license for a component/package

Fetch main license of the component/package from package-list/package repo, this will help the users to reach quick decide on the license situation based on the main license and license histogram of the package.

| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | - |
| Risk/Exploratory | - |
| Fun/Periphial | ** |
| Core Development | *** |
| Project Infrastructure | * |
| Project size | Large |
| Preferred contributor | Student/Professional |

**Reply by avinal** [reply]
This is a good one. I am wondering if we are aiming for another agent or incorporating this feature in any existing agent.

**Reply by GMishx** [reply]
It might not require a separate agent as the idea is to provide the package info and fetch the information from package manager APIs like [maven](https://central.sonatype.org/search/rest-api-guide/), [npm](https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md), etc. 
The idea is to fetch the main license from the package manager and allow clearing experts to make decision on clearing efforts based on license type (permissive, copyleft, etc.)

**Reply by ro4i7** [reply]
> It might not require a separate agent as the idea is to provide the package info and fetch the information from package manager APIs like [maven](https://central.sonatype.org/search/rest-api-guide/), [npm](https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md), etc. The idea is to fetch the main license from the package manager and allow clearing experts to make decision on clearing efforts based on license type (permissive, copyleft, etc.)

Hello @GMishx if this project is open.I am interested in this project

**Comment by ag4ums** [comment]
# Idea!
## Title: License as service
### Goal: License repository that can be used by other project/s

Idea is to create a license database similar to what fossology uses and expose it over an API which can then be used by other projects like fossology and sw360 as source of license text, risk, etc.

| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | - |
| Risk/Exploratory | - |
| Fun/Periphial | ** |
| Core Development | *** |
| Project Infrastructure | * |
| Project size | Large |
| Preferred contributor | Student/Professional |

**Reply by avinal** [reply]
Hi @ag4ums, did you mean, "a database for licenses"?

**Reply by ag4ums** [reply]
yes @avinal separate license database and the corresponding services/endpoint for licenses

**Reply by avinal** [reply]
Ok, nice idea, I am interested in this.

**Reply by viv9k** [reply]
I'm interested on this.

I have done some research similar mechanism is being used by clearly defined to store licenses.
We can this as proposed.

**Reply by sjha2048** [reply]
Seems Interesting, would love to help out students with this one, maybe we can combine the scope of this idea with [Obligation as service](https://github.com/fossology/fossology/discussions/2140#discussioncomment-2112774) as the underlying infra for both of the projects would be same, correct me if I am wrong @ag4ums.

**Reply by sanjaykrishna1203** [reply]
Hi @ag4ums sir, I am interested in this Project and would like to know more about the Tech stack we are going to use for this project. Can you please elaborate on that

**Reply by shaheemazmalmmd** [reply]
As part of development, contributor should combine both [license as a service](https://github.com/fossology/fossology/discussions/2371#discussioncomment-4774930) and [obligation as a service](https://github.com/fossology/fossology/discussions/2371#discussioncomment-4774943) as most of the code that is used to develop this two services will be same.

**Comment by ag4ums** [comment]
# Idea!
## Title: Obligation as service
### Goal: Obligation repository that can be used by other project/s

Idea is to create a obligations database similar to what fossology uses and expose it over an API which can then be used by other projects like fossology and sw360 as source of obligations for given license(s).

| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | - |
| Risk/Exploratory | - |
| Fun/Periphial | ** |
| Core Development | *** |
| Project Infrastructure | * |
| Project size | Large |
| Preferred contributor | Student/Professional |

**Reply by khanjasir90** [reply]
@ag4ums looks like a great idea would like to add value to this project. could u pls provide me points to get started with this.

**Comment by GMishx** [comment]
# Idea!
## Title: Improving FOSSology CI scanner image
### Goal: Enhancing current scanner image with new reports and features
As a fun project, FOSSology started combining scanners in a simple and small Docker image which can be run on CI providers. The image is currently capable of understanding build environment (GitLab/GitHub Actions/Travis) and use their API's to fetch diff of a branch or scan the complete repo. The capabilities of image include license scanning with Nomos and ojo scanners, copyright and keyword scanning with respective scanners. The image makes use of a Python script to perform all the tasks.

1. The script, however, currently prints report in text format on console and generate the same as an artifact. It can be improved by generating reports in other formats like SPDX reports. See [their Python API](https://github.com/spdx/tools-python).
2. The integration with GitHub Actions can be improved by reporting line number where a license violation is found.
3. Allowing user to provide a different list of Keywords for scanning (currently stored at `/usr/local/share/fossology/keyword/agent/keyword.conf`).
4. Improving on [whitelist format](https://github.com/fossology/fossology/wiki/FOSSology-scanners-in-CI#explanation) with feature to provide it from other sources, currently it is read from a file which is expected to be in the root of repo being scanned.
5. Ability to down a dependency for scan (path provided at pipeline trigger).

Additionally, the JSON output of nomos needs to be enhanced providing highlight and line information in the output.

| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | - |
| Risk/Exploratory | ** |
| Fun/Periphial | *** |
| Core Development | ** |
| Project Infrastructure | *** |
| Project size | Large |
| Preferred contributor | Student/Professional |

**Reply by vandyyyy** [reply]
looks like a great idea would like to add value to this project. could u pls provide me points to get started with this

**Comment by shaheemazmalmmd** [comment]
# Idea!
## Title: Support dynamic CSS class of bootstrap in FOSSology
### Goal: Create a PHP page and define all the CSS classes of bootstrap that are been used inside of project

FLOW :  Create a PHP page and define all the CSS classes of bootstrap that are been used inside of project.
        Also beautify the existing FOSSology UI with Bootstrap.
        We can easily change from versions of bootstrap if we have all the classes in one place.


##Discussion

- Search/Bucket/Refresh in Sub-menu really required ??
- Browse/ file Browser in tree view how may people use this ??
- Tag/Download/View in Tree view table required ??
- Spasht/software heritage required in sub-menu or move it to main menu ??


| Category | Rating |
| :-- | :-- |
| Low Hanging Fruit | ** |
| Risk/Exploratory | * |
| Fun/Peripheral | ** |
| Core Development | * |
| Project Infrastructure | ** |
| Project size | Large |
| Preferred contributor | Student |

**Reply by anshuman-dev** [reply]
Hi @shaheemazmalmmd , I am interested to be a mentor for this project.

**Reply by shaheemazmalmmd** [reply]
Hello @anshuman-dev Thank you for your interest in mentoring this project. Can you please join our community call to discuss this further. you can find the link for the call here https://github.com/fossology/fossology/wiki#contact-and-information

**Reply by anshuman-dev** [reply]
@shaheemazmalmmd . It was nice talking with you. Here's the link to my LinkedIn profile:
https://www.linkedin.com/in/anshuman-singh-4537731a5/

**Reply by Ayush-Kori** [reply]
Hi @shaheemazmalmmd . I am interested in contributing in the project.

**Reply by Simran180** [reply]
Hi @shaheemazmalmmd . I am interested to contribute to this project .

**Reply by PranitPatil03** [reply]
I am interested in contributing to this project and would like to learn more about the Project.

**Reply by sumit851** [reply]
i want to contribute to this project, i also know html css jquery bootstrap java it would help me to contribute to this project .as a beginner

**Reply by Nishant123455** [reply]
hello @shaheemazmalmmd I will love to give time for this Idea.  I want to contribute to this project

**Comment by pratham-yuxo** [comment]
Hello everyone,
I would love to work with your organization, I am experienced in MERN stack. So, can you please tell me what technologies should I learn to be helpful in your work and can you please assign me some good first issues.

**Comment by sahil5758** [comment]
Hey,I am sahil from Chandigarh. i am in 3rd year of  bachelors of techonology in cse from cgc college of engineering,landran.
i want to contribute in bootstrap project . can u guys guide me how to get started as i am beginner to open source and but i am willling to work with you guys to learn and apply my skills in the project

**Comment by shaheemazmalmmd** [comment]
Dear Contributors,


Thank you for showing interest in contributing to FOSSology.

We are hosting our first contributor call on 15th March 23 @ 2:30PM UTC to discuss about project ideas and answer any questions that contributors have.

Please use the below link to join the call.


[Click here to join the meeting](https://teams.microsoft.com/l/meetup-join/19%3ameeting_YTA1YWZjN2ItYTNlNi00ZGY1LWFhOTItYTJjYWI2N2ZlNWJk%40thread.v2/0?context=%7b%22Tid%22%3a%2238ae3bcd-9579-4fd4-adda-b42e1495d55a%22%2c%22Oid%22%3a%227f1979f2-e3e3-40d0-8a77-bcef842abf7e%22%7d)

Meeting ID: 350 787 404 800
Passcode: 8Xq3YT

**Reply by Chetan-KK** [reply]
hello i also want to contribute to this project, i'm little late but yes i have extensive knowledge on frontend development and i love to work on UI/UX development can you guide me some more about this.
here is my portfolio : https://chetan-kk.github.io/Chetan-KK/
